tsfc/driver.py

```diff
@@ -434,6 +370,55 @@ def compile_expression_dual_evaluation(expression, to_element, coordinates, *,
     return builder.construct_kernel(return_arg, impero_c, index_names)
 
 
+class UFLtoGEMCallback(object):
+    def __init__(self, expression, kernel_cfg, complex_mode):
+        # UFL expression to turn into GEM
+        self.expression = expression
+
+        # Geometric dimension or topological dimension
+        # For differentiating UFL Zero
+        # Fix for Zero expressions with no dimension
+        # TODO: Fix for topolgical dimension different from geometric
+        try:
+            self.dimension = expression.geometric_dimension()
+        except UFLException:
+            self.dimension = 0
```

Comment:
So this is only for when getting the derivative of a `ufl.ConstantValue` which should be a `ufl.constantvalue.Zero` of the right shape. Since you appear to be copying [these lines](https://github.com/FEniCS/ufl/blob/81d1217792e00a4af5014a2d458da94077773883/ufl/differentiation.py#L179-L187) from `ufl.differentiation.ReferenceGrad` shouldn't this query `expression.topological_dimension`? 

Furthermore, doesn't this fix, if required, belong in a UFL PR?

Reply:
Using `expression.ufl_domain().topological_dimension()` is indeed better, though it also fails for `test_constant` in `test_interpolate.py` since the `Constant` doesn't have a UFL domain or specified dimension. Any ways around it?

Doing this in UFL is probably the better solution, but @dham and I decided not to bother with doing a UFL PR due to time constraints earlier. Is now the time for it? With @rckirby pointwise dual branch on FIAT which can express dual vectors with derivatives as weighted sums of point evaluations only, `ReferenceGrad` might not even be needed.